src/SymmetricPositiveDefinite.jl

```diff
@@ -215,7 +215,7 @@ denotes the lower triangular matrix with the diagonal multiplied by $\frac{1}{2}
 function exp!(M::MetricManifold{SymmetricPositiveDefinite{N},LogCholeskyMetric}, y, x, v) where N
     (l,w) = spd_to_cholesky(x,v)
     z = exp!(CholeskySpace{N}(),y,l,w)
-    mul!(y,z,z')
+    y .= z*z'
```

Comment:
Are you sure this is necessary? I thought the old version was more memory efficient?

Reply:
In-place multiplication is `mul!`!
```julia
using LinearAlgebra, BenchmarkTools
A = randn(5,5)
S = similar(A)
fun!(z, y) = z .= y*y'
@btime fun!($S, $A);
@btime mul!($S, $A, $A');
```
The first one (`fun!`) is _not_ allocation-free.

Reply:
Well, `mul!(y,z,z')` gives wrong answer for `MMatrix`. I've reported it here: https://github.com/JuliaArrays/StaticArrays.jl/issues/697 . In this code `z` and `y` is the same matrix, `LinearAlgebra` handles it correctly but `StaticArrays` doesn't. Since one additional allocation won't slow this down too much, having a bit slower but correct implementation seems better until `StaticArrays` is fixed.

Reply:
In `mul!`, the storage variable must not be aliased with any of the factors, as the docstring clearly states. There can't be an in-place multiplication "while you're multiplying". So your issue over at `StaticArrays.jl` is not an issue: there is simply no guarantee of what happens when `C` is aliased with either `A` or `B`. What happens if the first argument is different from the second and third for `MMatrix`, as in `mul!(y, z, z')`?

Reply:
I didn't read that docstring and assumed that the `mul!` here is correct. It looks like `mul!` only accidentally works on `Array` here. The difference between `mul!` for `MMatrix` and `Matrix` only happens when there is aliasing. Here `y` and `z` are supposed to be the same matrix, it looks like introducing `z` is just misleading.

Reply:
Oh. Indeed. With the line 217, `z` is the same as `y`, we should avoid.

Reply:
Yes, as suggested in https://github.com/JuliaNLSolvers/Manifolds.jl/pull/27#discussion_r349890992, for exactly that reason.

Reply:
Oh! Sorry, I must have misread that (though I am not sure how that happened) but of course an `exp` on Cholesky helpes a lot in this case. Sorry for having missed that. One should definitely incorporate that comment here, then.

Reply:
OK, so I'll change it to a non-mutating `exp`.

Reply:
I've changed this so I think this PR can be merged once CI passes.

Reply:
Now you should write `mul!(y, z, z')` to have a truly non-allocating multiplication.

Reply:
Of course, I forgot about it. Thanks!

Reply:
Except it it doesn't work for `MMatrix` :/

Reply:
What? What's the type of `z`? Where does `lmul!` come from?

Reply:
Hm, I see. I'll double-check.

Reply:
This is really tedious... Seems like there is no triangular multiplication routines for `MMatrix`.

Reply:
Okay, it seems there is no in-place `mul!` for triangular `[S/M]Matrix`, only `*`, see https://github.com/JuliaArrays/StaticArrays.jl/blob/master/src/triangular.jl. So the line above should be okay for those types. It's suboptimal for `Base` arrays (and larger static matrices that fall back to BLAS), but maybe these are not much used in this context.

Reply:
OK,, we can improve this another time.